rfcs/rfc-colin-states-mapping-to-lear.md

```diff
@@ -0,0 +1,127 @@
+
+_Example of what this might look like but will need to represent sub-types of filings where applicable_
+``` http request
+
+GET https://legal-api-dev.apps.silver.devops.gov.bc.ca/api/v2/businesses/CP0001199
+
+{
+    "business": {
+        "legalName": "XYZ TEST CORP.",
+        "legalType": "BC",
+        "state": "HISTORICAL",
+        "currentlyActionableFilingTypes": ["putBackOn", "restoration"],
+        ...
+}
+```
+
+An invalid business state transition map will be implemented as a python dictionary or database table.  This map can be used in conjunction with the existing [ALLOWABLE_FILINGS](https://github.com/bcgov/lear/blob/e850f9a22672910db6e4ceb2f1ddb9437541a86f/legal-api/src/legal_api/services/authz.py#L96) dictionary as well as any other conditions to determine what filings are allowed for a business at a given point in time.
+
+_Example of what invalid state transition might look like_
+``` python
+INVALID_STATE_TRANSITION_FILINGS: Final = {
+    'staff': {
+        Business.State.ACTIVE: { # current business state
+            'restoration': { # target filing
+                # invalid target filing if filing that transitioned business to its current state is in list
+                'fullRestoration': ['registration', 'incorporationApplication'], 
+                'limitedRestoration': ['registration', 'incorporationApplication'], 
+            }
+            ...
+        },
+        Business.State.HISTORICAL: {
+            'restoration': { 
+                'fullRestoration': ['almagmation', 'continuationOut', 'continuationIn'], 
+                'limitedRestoration': ['almagmation', 'continuationOut', 'continuationIn'],
+            },
+            'putBackOn': ['almagmation', 'withdrawal']
+            ...
+        },
+        Business.State.LIQUIDATION: {
+            'restoration': {
+                'fullRestoration': ['liquidation'], 
+                'limitedRestoration': ['liquidation'], 
+            },
+            ...
+        }
+    },
+    'client': {
+        Business.State.ACTIVE: { # current business state
+            ...
+        },
+        Business.State.HISTORICAL: {
+            ...
+        },
+        Business.State.LIQUIDATION: {
+            'restoration': {
+                'fullRestoration': ['liquidation'], 
+                'limitedRestoration': ['liquidation'], 
+            },
+            'putBackOn': ['liquidation']
+            ...
+        }
+    }        
+    
+}
+```
+
+_Allowable state transitions matrix in COLIN for restoration and put back on filings_
+![colin_allowable_filings.png](rfc-colin-states-mapping-to-lear/colin_allowable_filings.png)
+
+Filing validation logic will be updated to include logic to ensure that only valid business state transitions are allowed. 
+
+All filings resulting in business state transitions will need to populate the businesses.state_filing_id field.  This is used to determine if future business state transitions are permitted.
```

Comment:
I know we already use the state filing for VDs. Do we only need to keep track of the last filing that changed state?

Reply:
In the BE we already have function to check the latest filing for a business.  Might not be exposed by the API though via the business response though.

But for the purposes of being able to determining what filings are allowed for a business at a given point in time, `businesses.state_filing_id` is used in conjunction with the ALLOWABLE_FILINGS dictionary.